ecosystem/gstreamer_plugin/gst_mtl_common.c

```diff
@@ -422,5 +456,50 @@ mtl_handle gst_mtl_common_init_handle(struct mtl_init_params* p, StDevArgs* devA
   }
   *log_level = mtl_init_params.log_level;
 
-  return mtl_init(&mtl_init_params);
+  if (force_to_initialize_new_instance) {
+    GST_INFO("MTL shared handle ignored");
+    pthread_mutex_unlock(&common_handle.mutex);
+
+    return mtl_init(&mtl_init_params);
+  }
+
+  common_handle.mtl_handle = mtl_init(&mtl_init_params);
+  common_handle.mtl_handle_reference_count++;
+  pthread_mutex_unlock(&common_handle.mutex);
+
+  return common_handle.mtl_handle;
+}
+
+/**
+ * Deinitializes the MTL handle.
+ * If the handle is the shared handle, the reference count is decremented.
+ * If the reference count reaches zero, the handle is deinitialized.
+ * If the handle is not the shared handle, it is deinitialized immediately.
+ *
+ * @param handle MTL handle to deinitialize (Null is an akceptable value then
+ * shared value will be used).
+ */
+gint gst_mtl_common_deinit_handle(mtl_handle handle) {
+  int ret;
+
+  pthread_mutex_lock(&common_handle.mutex);
+
+  if (handle && handle != common_handle.mtl_handle) {
+    pthread_mutex_unlock(&common_handle.mutex);
+    return mtl_uninit(handle);
```

Comment:
again, I'm not sure if we can call this in parallel. call it under lock or even better prove me wrong

Reply:
Done